### ISSUE TITLE
2.0 design

### ISSUE THREAD

**Issue by yyx990803** [vuejs/vuex#236]
> Note: this proposal has been heavily revised based on feedback so some of the comments below may be out of context.

### 1. Terms naming change for better semantics

Dispatching a mutation never sounded right. Dispatch should be indicating the intention for something to happen. For mutations, we want a verb that indicates the state change is happening as soon as you call it.

- The old `store.dispatch` is now `store.commit`.
- `store.dispatch` will be used for firing actions instead. (see next section)

The new naming better conveys the semantics behind the two methods:

- "dispatching an action": indicating the intention for something to happen (possibly async with side effects).
- "committing a mutation": indicating the synchronous transaction of actual state change.

#### Actions inside store and modules

There has been common requests on shipping actions with the store, or inside modules. Previously, the reason for not putting actions in the store/modules was mainly how do we access them. Actions defined inside the store means we need access to the store - again the singleton problem, which is now a non-issue.

Now since they are just functions, we may expose them as `store.actions` and call them directly. In fact this was the API of Vuex 0.4.x:

``` js
store.xxx // reserved for Store class methods/properties
store.actions.xxx // this seems ok
```

A problem is when actions are defined inside modules, what if multiple modules define actions of the same name? Also, sometimes we may want to call an action that affect multiple stores, just like mutations.

It seems actions are also more like event listeners. Instead of calling them directly as functions, we now use `store.dispatch` to trigger them:

``` js
const store = new Vuex.Store({
  actions: {
    doSomething: ({ commit }) => {
      commit('some-mutation')
    }
  }
})

store.dispatch('doSomething')
```

This way, you can dispatch actions in multiple modules with a single call, just like mutations. It's also more explicit that you are firing off some side-effects in your store, instead of just calling a random function.

#### Getters, too

You can now define getters in the store / modules too. Similar to module mutations, module getters receive the sub state tree:

``` js
const store = new Vuex.Store({
  state: {
    count: 0
  },
  getters: {
    hasAny: state => state.count > 0
  }
})

// access the getter
store.getters.hasAny // -> false
```

### 3. Composable Action Flow

Now that we are putting actions inside modules and calling them via `dispatch`, we somehow lose out on the composability of actions, because they are no longer just functions you can call. Here's how we can make them composable again:

1. To indicate the completion of an action, return a Promise from the action. `store.dispatch` will return that Promise if there is only a single handler called. If multiple action handlers are matched, it will return a Promise that resolves when all Promises returned by those handlers are resolved.

  ``` js
  const store = new Vuex.Store({
    actions: {
      doSomething: ({ commit }, payload) => {
        return callPromiseAPI(payload).then(res => {
           commit('some-mutation', { res })
        })
      }
    }
  })

  store.dispatch('doSomething', { id: 1 }).then(() => {
    // action done
  })
  ```

2. Based on (1) and `async/await`, we can have very clean composition between async actions:

  ``` js
  const store = new Vuex.Store({
    actions: {
      one: async ({ commit }, payload) => {
        const res = await callPromiseAPI(payload)
        commit('some-mutation', { res })
      },
      two: async ({ dispatch, commit }) => {
        await dispatch('one')
        commit('done')
      }
    }
  })

  store.dispatch('two') // fires off complicated async flow
  ```

  The convention of returning Promises also allows Vuex to:

  1. better handle errors during async action flow.

  2. simplify store initialization during server-side rendering.

### Component Binding

The design of Vuex 0.6~1.0 had a somewhat annoying design constraint: avoid directly accessing stores inside components. The store is injected at the root component, and implicitly used via the `vuex: { getters, actions }` options. This was in preparation for Vue 2.0 SSR (server-side rendering), because in common SSR setups (directly requiring the component in Node.js and render it), dependence on a global singleton will cause that singleton to be shared across multiple requests, thus making it possible for a request to pollute the state of the next one.

However, with the new `bundleRenderer` strategy implemented in Vue 2.0.0-alpha.7, this is no longer an issue. The application bundle will be run in a new context for each request, making it unnecessary to structure your app without singletons just for the sake SSR. This also means it's totally fine to just `import store from './store'`, use plain computed properties to return `store.state.xxx`, or calling `store.dispatch()` in plain methods.

This opens up path to simplifying the component binding usage, since theoretically you don't need any binding at all. Currently, the `vuex` options feels a bit clumsy and indirect.

In Vuex 2.0, the `vuex` option will be deprecated in favor of just computed properties and methods. You are free to structure your Vuex store usage the way you prefer. However, we will be keeping the injection for `this.$store` so that you can do this:

``` js
export default {
  computed: {
    a () {
      return this.$store.getters.a
    }
  },
  methods: {
    b (...args) {
      this.$store.dispatch('b', …args)
    }
  }
}
```

The above alleviates the need to import the store everywhere. But it can get verbose when you have many getters and actions in the same component. Therefore we provide two helpers, `mapGetters` and `mapActions`:

``` js
import { mapGetters, mapActions } from 'vuex'

export default {
  computed: mapGetters(['a', 'b', 'c']),
  methods: mapActions(['d', 'e', 'f'])
}
```

So in the component, `this.a` maps to `this.$store.getters.a`, and `this.d(...args)` maps to `this.$store.dispatch('d', ...args)`.

If you want to map a getter/action to a different local name, use an object instead:

``` js
import { mapGetters, mapActions } from 'vuex'

export default {
  computed: mapGetters({
    myComputed: 'a' // map this.myComputed to store.getters.a
  }),
  methods: mapActions({
    myMethod: 'b' // map this.myMethod() to store.dispatch('b')
  })
}
```

Finally, you can easily compose them with local computed properties and methods using Object spread operator:

``` js
import { mapGetters, mapActions } from 'vuex'

export default {
  computed: {
    localComputed () { … },
    ...mapGetters(['a', 'b', 'c', 'd'])
  },
  methods: {
    localMethod () { … },
    ...mapActions(['b'])
  }
}
```

### 2.0 so soon? How about 1.0?

1.0 contains small breaking changes but should be a very easy upgrade for existing 0.6~0.8 users. It will be maintained as a stable release in parallel to 2.0.

**Comment by NicolasParada** [comment]
My thoughts are on the order of params inside actions.
I mean:
```js
const store = new Vuex.Store({
  actions: {
    doSomething(state, payload, dispatch) {
      dispatch('anAction', payload)
    }
  }
})
```
...could be the in the inverse order:
```js
const store = new Vuex.Store({
  actions: {
    doSomething(dispatch, payload) {
      dispatch('anAction', payload)
    }
  }
})
```
Because, inside an action you always use `dispatch`, then, if there is a `payload`, you use it.

**Comment by yyx990803** [comment]
@NicolasParada definitely, the argument orders are not set in stone - I actually was thinking about it too.

**Comment by sleewoo** [comment]
how about dynamic modules injection?
like store start with a basic set of modules then user can fetch and inject new ones.
and presumably delete them when no more needed.

**Comment by sebastiandedeyne** [comment]
> This opens up path to further simplifying the component binding usage. Currently I'm interested to hear thoughts on the current binding (aka vuex: { getters, actions }).

This might have been discussed at some point, but is the `vuex` key necessary? I'd prefer to  declare `getters` and `actions` on the top level of my component.

> `store.dispatch('counter:incrementAsync')`

I'm not sure about namespacing actions. Wouldn't that restrict you from using modules if you want a single action that can trigger multiple mutations?

**Comment by blake-newman** [comment]
Mainly, I agree with all the concepts and ideas brought in by this design. It will certainly help further simplify Vuex.

> This opens up path to further simplifying the component binding usage. Currently I'm interested to hear thoughts on the current binding (aka vuex: { getters, actions }).

The `vuex` property key, in the current architecture is kind of necessary. For 2.0 I would consider it very unnecessary:

- Computed properties and methods are learn't in Vue.
 -  Current Vuex we are adding another abstraction which is essentially thesamee thing as methods and computed properties.
 - This adds an additional learning curve.

> Actions and mutations are now together "events". The difference being that 
- mutations are synchronous and can mutate the state
- actions can be async, cannot directly mutate the state (but can trigger mutations and other actions)

I would replace `store.dispatch` with equivalent `store.action` and `store.mutate`.

- Helps isolate the concepts through the API.
- Learning Vuex includes have to learn the flux principles, therefore `dispatch` becomes something additional to digest that does not fit in with what is learnt through Flux. 
- The naming `mutate` and `action` fits nicely into the concepts learnt from Flux.
- This may also help with slight performance gains as we are splitting of the target for the events

In agreement with @NicolasParada, this defiantly makes simple actions easier to create.

**Comment by ruiposse** [comment]
I agree with @blake-newman that `store.action` and `store.mutate` is better than `store.dispatch`.
I would also add that is more readable because we can immediately understand what happening on the store.

**Comment by simplesmiler** [comment]
Interesting, quite a step away from Redux.

Redux community tried to [bundle actions with reducers](https://github.com/erikras/ducks-modular-redux), but I believe Dan is against it. According to him, it kills the essence of flux: action objects and reducers are separated, any reducer can respond to any action object. That said, bundled modules are optional, so you can use then when they fit and fall back to idiomatic Redux when they don't.

To be honest, I don't like the idea of dispatching actions and mutations through a single method. I think that everything that is dispatched should be recorded by the devtools. But actions should not be recorded because of side effects. In Redux you can dispatch anything, and "non-idiomatic" chunks are processed by middleware. But it's quite confusing, I'd rather have separate methods.

---

Onto suggestions.

With `dispatch` now performing prefixing (and thus being unable to affect parent modules), how about making actions take a `module`, which has it's local versions of `module.dispatch` and `module.state`? This would simplify action signature from `state, payload, dispatch, [done]` to `module, payload, [done]`. We can even keep calling it `store`, but it will change semantincs to be a "local store, aka module".

**Comment by yyx990803** [comment]
@simplesmiler I think there is a real need for actions defined alongside mutations, which doesn't prevent a higher-up action from dispatching multiple namespaced mutations.

I like the arguments simplification - it also keeps it closer to the current API, so the signature remains almost he same:

- mutations: `(state, payload) => void`
- actions: `({ dispatch, state }, payload, [done]) => any`

**Comment by yyx990803** [comment]
@sleewoo that's something I've been thinking too! It could also be used for hot-reloading a module. Maybe something like `store.module(id, { ... })`.

**Comment by yyx990803** [comment]
@sebastiandedeyne I also feel the nesting is a bit awkward. In 2.0 we may get rid of the options all together, so you can just do `this.$store.dispatch()` in normal methods + `return this.$store.state.xxx` in computed properties.

If you want a single action to trigger multiple mutations, you just need to be more explicit:

``` js
dispatch('moduleA:increment')
dispatch('moduleB:increment')
```

I think this actually makes the data flow clearer: you know you are triggering mutations in two different modules and what modules are affected.

We may also need an "escape hatch" for dispatching mutations from root scope: `dispatch('@rootModule:mutation')`

**Comment by yyx990803** [comment]
As for separating methods for mutations and actions - this is also something I thought about. Maybe we keep `dispatch` (since it's already been used with mutations) and use `call` for actions. `action` itself is not a verb so it feels weird. `call` seems to signify the "side-effect-ness" of actions pretty well.

**Comment by sebastiandedeyne** [comment]
Feels kind of weird though, all other libraries teach us to dispatch actions, and then Vuex would call actions and dispatch mutations.

I agree with the separation, just not with the naming.

**Comment by simplesmiler** [comment]
> We may also need an "escape hatch" for dispatching mutations in root scope from any module:
>
> `dispatch('@someRootMutation')`
> `dispatch('@rootModule:mutation')`

That would be nice. Maybe even consider them global (or application-wide) instead of root-scoped, and let any module dispatch and react to `@global` mutations.

**Comment by simplesmiler** [comment]
Also, should new Vuex encourage `state: () => <state>` instead of `state: <state>`, now that modules will be essentially reusable?

**Comment by OEvgeny** [comment]
Keeping actions to work with only one module/mutation sounds a bit wired to me. Imagine I need to implement some module to collect stats for my app. With redux I can just attach reducer to my store and make it respond to actions I need. With Vuex I forced to be more explicit and duplicate `dispatch` calls of needed actions to perform additional mutations.

**Comment by ktsn** [comment]
I think the design concept is very reasonable. In my developments, I always prefix or suffix mutation names for a related module. (and also I do it on Redux)

And I also agree with @simplesmiler. Explicitly dispatching multiple mutations is good, but I think sometimes we want to dispatch some mutations globally.
For example, imagine an app that can search some complex data and show various visualization on one page. I want to reset all state of previous search for each search query submission.

In my opinion, global registration would be better than global dispatch because global dispatch can accidentally execute unexpected mutations.

```js
// normal mutation name
const RESET = 'reset'

// global mutation name
// e.g. prefixed with special character is global
const GLOBAL_RESET = '#reset'

const someModule = {
  state: { value: '' },
  mutations: {
    [RESET] (state) { // register under the namespace
      state.value = ''
    }

    [GLOBAL_RESET] (state) { // register to global
      state.value = ''
    }
  }
}

// RESET is namespaced
store.dispatch(`someModule:${RESET}`)

// GLOBAL_RESET is not namespaced
store.dispatch(GLOBAL_RESET)
```

**Comment by kazupon** [comment]
I basically agree all the concept and ideas and @blake-newman. And I think that we should be provided the useful Vuex API I/F for user.

**Comment by yyx990803** [comment]
@OEvgeny even in Redux it feels weird using a reducer for collecting app analytics - a middleware seems more appropriate. So in Vuex you'd just write a plugin for that purpose.

**Comment by OEvgeny** [comment]
@yyx990803 thanks for response. I think there are different types of stats that can be collected. Some of them we don't need to process on the server side, some we need to collect before processing, and some we will use to show in featured parts of our app.

I like what with redux we can continue changing our store logic without touching whole app. As an example we can rewrite some parts of our store to fit new needs and implement new components/views for our new state without breaking or modifying our actions and current components/views. We can mix up together new and old parts of our store until it's needed. So, it gives more flexibility for further development.

**Comment by yyx990803** [comment]
@OEvgeny I don't think namespaced modules would make it any harder - I'm not sure what aspects in the proposal makes you think that way?

**Comment by OEvgeny** [comment]
@yyx990803 I think it's all about forcing mutations inside modules to be namespaced. Then each time when I change structure of my store, I will be forced to change actions to fit this changes. I think it can be hard to do in some cases.
As an example redux also gives the ability to namespace actions without forcing to do it. It allows developer to decide not only how store organized but how actions will affect to it.

**Comment by yyx990803** [comment]
1. Each time you change the structure of your redux store, you will also
need to change all the mapStateToProps connections that are affected. Also
any action creators that uses getState() are affected too. If your state
shape changed, you always have some work to do.

2. Namespacing is not forced. If you want you can still use global
mutations inside modules, which is basically the same behavior as now.

3. The problem with not (supporting) namespacing it, is that it's awkward
to write portable bundles of reducers + action creators. (in Vuex term,
mutations + actions) If an action creator relies on knowledge of the
current state, it has no way to know what path the coupled reducer is
nested under. Essentially, all redux action creators and middlewares are
forced to be global. (The only way around this is creating the action
creators with a path resolver when you combine the reducers, which also
makes it coupled to the structure of your store).

**Comment by simplesmiler** [comment]
With an **option** to operate on global scope you have my thumbs up.

**Comment by OEvgeny** [comment]
@yyx990803 
1. Yes you are right. Usually we need to change ``mapStateToProps`` and some actions which use  ``getState()``. But there are cases when it is not needed. For example we can add new reducer to our state and get it to work with existing actions while we are working on some new features/components and migrate existing components which rely on "old" part of state later.
2. Good to know. It was not clear from first message:

>## Mutations inside modules become namespaced

I'm agree with @simplesmiler about some option to achieve global actions behavior.

**Comment by yyx990803** [comment]
Actually, now that I think about it, there's no need for auto-namespacing, because users can namespace the mutations/actions themselves. This does leave the potential for namespace clashes, but the chance seems quite low. All we really need is auto state-resolving for module actions.

**Comment by yyx990803** [comment]
For all those interested, the proposal has been heavily revised. Feedback welcome.

**Comment by piknik** [comment]
I like the use of `mapGetters` and `mapActions`, it's what I've originally thought of! Glad to see it being mentioned to assist with development.

The terms seem reasonable to me and I invite the change, but others including those coming from Redux may be unfamiliar. I wouldn't care if they were unchanged, there are much worse terms given for certain concepts like *cough* **OpenGL** *cough*, and development still continues on.

The most important change in this proposal would be the modular getters and actions, for certain. Really helps in big projects. The composable actions are very nice with the promise support. I like this proposal!

**Comment by yyx990803** [comment]
@sleewoo that kind of implicitness is too much for an optional library like Vuex. When you see `mapGetters` and `mapActions`, the intentions are very clear because you see that they are imported from `vuex`.

**Comment by sleewoo** [comment]
makes sense.
just "realised" i treat vuex as the main piece in the stack, others being just "effects" :)

**Comment by theotherzach** [comment]
I'd for our designers / frontend developers / template authors to be able to declare getters much like we do props. 

```html
<my-component v-for="item in items" :item="item" v-map-getters(['a', 'b', 'c'])>
</my-component>
```

**Comment by yyx990803** [comment]
@theotherzach you can just do `:a="$store.getters.a"`.

**Comment by pdcmoreira** [comment]
Can we use Vuex 2.0 with Vue 1.0?

**Comment by ktsn** [comment]
@pdcmoreira Yes, Vuex 2.0 works with Vue 1.0

**Comment by sebastiandedeyne** [comment]
Would a `mapMutations` function make sense? I'm looking to reduce the boilerplate of creating actions from mutations. I'm currently upgrading an existing app to 2.0 and I have a lot of actions that just pass through to a mutation.

Example:
```js
const actions = {

    restoreMedia({ commit }, { media }) {
        commit('restoreMedia', { media });
    },

    clearCollection({ commit }, { collection }) {
        commit('clearCollection', { collection });
    },

    setMediaOrder({ commit }, { order }) {
        commit('setMediaOrder', { order });
    },

    updateCustomProperty({ commit }, { media, prop, value }) {
        commit('updateCustomProperty', { media, prop, value });
    },

};
```

Alternatively, alt.js has a `generateActions` function for cases like this, maybe something similar would be handy?

http://alt.js.org/docs/createActions/

**Comment by yyx990803** [comment]
@sebastiandedeyne yep, this is really trivial to implement: https://github.com/vuejs/vuex/blob/next/examples/todomvc/vuex/actions.js#L8-L10 although it might be a good idea to ship it so you can just import it as needed

**Comment by theotherzach** [comment]
The boilerplate for the simple case isn't great, that's for sure. Vue alone for single components or components nested 1 deep feels amazing. Vuex for complex shared actions that commit multiple mutations, also great.  

What isn't great is when I just need to share simple data across large-ish app, especially if I got that data from the server, especially if I'm providing optimistic updates and potentially server error validations for multiple fields. I'm constantly making the "wrong" decision about where stuff should go with regards to individual components vs actions vs mutations.

**Comment by yyx990803** [comment]
@theotherzach not quite understanding your second paragraph. If the data is shared by multiple components then it should obviously be in the store. As for actions vs mutations, the rule is pretty simple: the only thing mutations do is changing the state. It gets a value, sets the value, that's it. Anything else that is related to the store goes in to actions.

**Comment by theotherzach** [comment]
I'll write up one of my more painful experiences involving forms in a day or two. I'm currently prepping my Vue.js talk at tomorrow's Chicago HTML 5 meetup.

**Comment by sebastiandedeyne** [comment]
> @sebastiandedeyne yep, this is really trivial to implement: https://github.com/vuejs/vuex/blob/next/examples/todomvc/vuex/actions.js#L8-L10 although it might be a good idea to ship it so you can just import it as needed

In case anyone's interested, I also added a function to generate actions from all mutations in my application.

```js
const makeAction = (type) => (
    ({ commit }, payload) => commit(type, payload)
);

const makeActions = mutations => (
    Object.keys(mutations).reduce((actions, type) => {
        actions[type] = makeAction(type);
        return actions;
    }, {})
);

const mutations = { /* ... */ };

const actions = {
    ...makeActions(mutations),
    /* Custom actions... */
};
```

**Comment by karol-f** [comment]
Thank you @yyx990803 for Vuex **2.0** changes. Especially for dynamically registering nested modules into a store - I use Vue instances a lot, as a widgets, using modified [vue-element](https://github.com/vuejs/vue-element). 

I would like to propose some changes/improvements so we can auto register nested store modules in Vue component instances.

Currently, using Vuex **2.0**, my implementation looks like this:
**http://www.webpackbin.com/EkAdO7DUW**

If you will look at `widget.js` you will see that on `created()` I manually trigger `this.$store.module` with `this._uid` name.

I would like to propose further simplifying this case. There is an example in `widget.pseudocode.js` file, showing how it might look like:
* `storeModule` is auto registered in store when creating Vue instance
* name of nested store module is auto set using `this._uid`

I would love to here Your opinion on this. Keep up great work, regards!

**Comment by xhh2a** [comment]
Disclaimer: I don't use vuex and I am primarily a python programmer who's playing around with vue.js right now, maybe I'm just being an idiot here.

I've read a bit about the advantages of veux (eg http://skyronic.com/2016/01/03/vuex-basics-tutorial/) and I've looked at the 1.0 veux documentation. Is there a reason why there isn't a way to just define a setter in veux and access elements of the store similar to how ```@property``` works in python?

e.g. from the current documentation:
```
import Vuex from 'vuex'

const store = new Vuex.Store({
  state: {
    count: 1
  },
  mutations: {
    INCREMENT (state) {
      // mutate state
      state.count++
    }
  }
})
//Usage
store.dispatch('INCREMENT')  
```

To something like:
```
import Vuex from 'vuex'

const store = new Vuex.Store({
  state: {
    count: {
       get: function()
       set: function(val)
    }
  }
})

//Usage:
store.count = 5
//Or maybe store.count.set(5) and make it possible to allow other 'mutations' to be defined.
```
It'd be useful when you just want to do something simple with forms and you do want the state to be tracked by a central state. If anything, I find that more often than not (at least in python code) I don't need custom get/set logic so if veux could provide a default get/set that would reduce boilerplate considerably.

**Comment by theotherzach** [comment]
>not quite understanding your second paragraph. If the data is shared by multiple components then it should obviously be in the store.
@yyx990803

First of all, thank you so much for your work on Vue and its plugins. I consider Vue to be one of the few sane corners of the frontend JavaScript world which is why I promote it as much as possible.


Let's look at a component's method that I'm in the process of moving to an action, multiple mutations, and multiple getters. I've added comments to give additional context, but otherwise the code is pretty much as-is from work.

``` javascript
//snip
methods: 
  
  // triggered from a form submit
  registerUser: function () {
    var self = this
    var payload = {
      email: self.userEmail
    }
    // turns on loading indicator on our "Proceed to shipping" button
    self.isLoadingUser = true 
   // closes email form
    self.isEmailFormOpen = false
    app.services.usersService.save(payload).then(function (data) {
      // Successfully registered guest
  
      // turns off loading indicator
      self.isLoadingUser = false
      // calls an action that sets up the state of our checkout flow
      // to handle a guest checkout with the given email address.
      self.updateGuestEmail(payload.email)
    }, function (err)
      // Server error - possibly because the user is a returning guest

      // turn off loading indicator
      self.isLoadingUser = false
      var isUserRegistered = err.status === 409
      if (isUserRegistered) {
        // looks like this is a returning user, display the login form
        return self.isLoggedOutUser = true
    }

    // Open the email form back up
    self.isEmailFormOpen = true

    // Highlight the email field and display the server validation failure
    self.ajaxErrors = err.responseJSON

    // Call an action that flashes the appropriate message
    self.ajaxError(null, err.responseJSON, ["registerUser", "Register User"])
  })
},
```

Once I extract this method out of the component and into a single action, I'll need create many mutations and getters to get the data back to my components. Let's look at just this single line in my method:

```javascript
 // turns on loading indicator on our "Proceed to shipping" button
self.isLoadingUser = false
```

When in a component, dealing with that loading indicator is just one line. How many lines do I need after the method becomes an action?

```javascript
// 1. in the action
store.dispatch("LOADING_USER")
```

```javascript
// 2. in my state object
isLoadingUser: false,
```

```javascript
// our count is at 5 when I add 3 more from my mutations
LOADING_USER: function (state) {
  state.isLoadingUser = true
},
```

```javascript
// up to 8 lines when I add 3 more from my mutations
isLoadingUser: function (state) {
  return state.isLoadingUser
},
```

We're up to 8 lines of code to replace `self.isLoadingUser = false`. Let's call it 7 because I don't need `isLoadingUser` in my component's data object anymore, but that still feels like way too much work for a boolean value that I'm toggling.* 

```javascript
self.isEmailFormOpen = false
```

That's another 1 line that becomes 7 lines. 

```javascript
self.isLoggedOutUser = true
```

7 more lines.

```javascript
self.ajaxErrors = err.responseJSON
```

Let's call this 5 more lines because I can handle this specific case with a getter that filters my ajax errors.

The net result is that actions often feel far more painful than comparable component methods, so I try to keep as many of these as methods as possible. Unfortunately this bites me because the moment I extract something like `registerUser ` then any other methods that are setting extracted state such as `isLoadingUser` will no longer function. I either need to make an explicit action to set `isLoadingUser` (3 more lines) or extract those methods into actions as well. 

In summary, I really like the idea of Vuex, but I see the following as painful when I use it.

* A 7 factor increase in code weight for simple cases when compared to single component state.
* The network effect of action extraction. If you extract a method into an action, then any other methods that were relying on setting the now extracted state will break unless you also extract those methods or you create a specific action for setting that previously shared state.




*Sure, I could set up something that hooks into my ajax calls that automatically handles loading indicators, but I don't have a single site wide indicator, but indicators that are right near where the user is entering input, and I often need some custom handling especially in error cases, so I've not been able to figure out an automated way of handling it yet.

**Comment by theotherzach** [comment]
This is my 255 line getter file for this project, just for transparency. I probably shouldn't throw this in here, but whatever. It'll be in web browsers soon enough anyway.

```javascript
;(function () {
  "use strict"

  var app = window.app

  var getters = {

    /**
     * Returns either a logged in users email address or the email address
     * attached to an order.
     * @param {Object} state
     * @returns {(string|undefined)} - email address
     */
    email: function (state) {
      if (state.order.email) {
        return state.order.email
      }

      return state.user.email
    },

    creditCard: function (state) {
      return state.creditCard
    },

    COUNTRIES: function (state) {
      return state.COUNTRIES
    },

    errors: function (state) {
      return state.errors
    },

    /**
     * @returns {Boolean}
     */
    hasEmailAddress: function (state) {
      return app.getters.email(state) != undefined
    },

    hasEnteredNewCard: function (state) {
      return !(state.creditCard.last4 === undefined || state.creditCard.last4 === null)
    },

    /*
     * @returns {Boolean}
     */
    hasAttachedPayment: function (state) {
      if (app.getters.hasEnteredNewCard(state)) {
        return true
      }

      return state.reusedCardId !== null
    },

    hasSelectedInStore: function (state) {
      return !(state.pickupDetails.store_location_id === undefined || state.pickupDetails.store_location_id === null)
    },

    hasAttachedShippingAddress: function (state) {
      return !(state.order.shipping_address_id === undefined || state.order.shipping_address_id === null)
    },

    isAddressFormValid: function (state) {
      return state.isAddressFormValid
    },

    isCreditCardFormVisible: function (state) {
      if (state.hasActivatedEditCreditCard !== null) {
        return state.hasActivatedEditCreditCard
      }

      if (app.getters.paymentMethods(state)[0]) {
        return false
      }

      return !app.getters.hasAttachedPayment(state)
    },

    isCreditCardFormValid: function (state) {
      return state.isCreditCardFormValid
    },

    isEmptyCart: function (state) {
      return state.order.id != undefined && (!state.order.order_line_items ||
                                             !state.order.order_line_items[0])
    },

    isBillingDisplayed: function (state) {
      return app.getters.isShippingDisplayed(state) &&
        (state.order.shipping_method ||
         app.getters.hasSelectedInStore(state))
    },

    /**
     * @param {state}
     * @returns {boolean}
     */
    isEditingEmail: function (state) {
      return state.isEditingEmail
    },

    isEditingAddress: function (state) {
      return state.isEditingAddress
    },

    /** 
     * If a user's email address is found, but the are not 
     * currently logged in.
     * @param {Object} state
     * @returns {boolean}
     */
    isLoggedOutUser: function (state) {
      return state.isLoggedOutUser
    },

    /** 
     * If we're waiting on the server to return user info.
     * @param {Object} state
     * @returns {boolean}
     */
    isLoadingUser: function (state) {
      return state.isLoadingUser
    },


    isOnlySelectingInStore: function (state) {
      return state.isOnlySelectingInStore
    },

    isOrderLoading: function (state) {
      return state.order.id === undefined || state.order.id === null
    },

    isReusingAddress: function (state) {
      return state.reusedAddressId !== null
    },

    isReviewDisplayed: function (state) {
      return app.getters.isBillingDisplayed(state) &&
        app.getters.hasAttachedPayment(state)
    },

    isReviewOrderButtonDisplayed: function (state) {
      if (app.getters.hasAttachedPayment(state)) {
        return false
      }

      return true
    },

    /**
     * @param {Object} store
     * @returns {Boolean}
     */
    isShippingMethodSelected: function (state) {
      return state.order.shipping_method !== ""
    },

    isShippingDisplayed: function (state) {
      return !app.getters.isOrderLoading(state) && app.getters.hasEmailAddress(state)
    },

    isUsingNewAddress: function (state) {
      if (!app.getters.hasAttachedShippingAddress(state)) {
        return false
      }

      return !app.getters.isReusingAddress(state)
    },

    isWaitingForFinish: function (state) {
      return state.isWaitingForFinish
    },

    order: function (state) {
      return state.order
    },

    paymentMethods: function (state) {
      return state.paymentMethods
    },

    pickupDetails: function (state) {
      return state.pickupDetails
    },

    /**
     * When the server returns a credit card error on finalize,
     * and the user is attempting to reuse a card, that UserCard's
     * id is pushed into this array.
     * @param {Object} state
     * @returns {Array<Number>}
     */
    problemCardIds: function (state) {
      return state.problemCardIds
    },

    reusedAddressId: function (state) {
      return state.reusedAddressId
    },

    reusedCardId: function (state) {
      return state.reusedCardId
    },

    shippingAddress: function (state) {
      return state.shippingAddress
    },

    shippingContact: function (state) {
      return state.shippingContact
    },

    shippingMethods: function (state) {
      return state.shippingMethods
    },

    shippingPromoDollars: function (state) {
      var toDollars = Vue.filter("toDollars")
      return toDollars(state.shippingPromoCents)
    },

    STATES: function (state) {
      return state.STATES
    },

    storeLocations: function (state) {
      return state.storeLocations
    },

    toQualifyDollars: function (state) {
      var toDollars = Vue.filter("toDollars")
      var difference = toDollars(state.shippingPromoCents - state.order.subtotal_cents)

      if (difference <= 0) {
        return 0
      }

      return difference
    },

    user: function (state) {
      return state.user
    },

    userAddresses: function (state) {
      return state.userAddresses
    }

  }

  app.getters = getters

})();
```

**Comment by theotherzach** [comment]
```javascript
;(function () {
  "use strict"

  var app = window.app

  var getters = {

    /**
     * Returns either a logged in users email address or the email address
     * attached to an order.
     * @param {Object} state
     * @returns {(string|undefined)} - email address
     */
    email: function (state) {
      if (state.order.email) {
        return state.order.email
      }

      return state.user.email
    },

    creditCard: function (state) {
      return state.creditCard
    },

    COUNTRIES: function (state) {
      return state.COUNTRIES
    },

    errors: function (state) {
      return state.errors
    },

    /**
     * @returns {Boolean}
     */
    hasEmailAddress: function (state) {
      return app.getters.email(state) != undefined
    },

    hasEnteredNewCard: function (state) {
      return !(state.creditCard.last4 === undefined || state.creditCard.last4 === null)
    },

    /*
     * @returns {Boolean}
     */
    hasAttachedPayment: function (state) {
      if (app.getters.hasEnteredNewCard(state)) {
        return true
      }

      return state.reusedCardId !== null
    },

    hasSelectedInStore: function (state) {
      return !(state.pickupDetails.store_location_id === undefined || state.pickupDetails.store_location_id === null)
    },

    hasAttachedShippingAddress: function (state) {
      return !(state.order.shipping_address_id === undefined || state.order.shipping_address_id === null)
    },

    isAddressFormValid: function (state) {
      return state.isAddressFormValid
    },

    isCreditCardFormVisible: function (state) {
      if (state.hasActivatedEditCreditCard !== null) {
        return state.hasActivatedEditCreditCard
      }

      if (app.getters.paymentMethods(state)[0]) {
        return false
      }

      return !app.getters.hasAttachedPayment(state)
    },

    isCreditCardFormValid: function (state) {
      return state.isCreditCardFormValid
    },

    isEmptyCart: function (state) {
      return state.order.id != undefined && (!state.order.order_line_items ||
                                             !state.order.order_line_items[0])
    },

    isBillingDisplayed: function (state) {
      return app.getters.isShippingDisplayed(state) &&
        (state.order.shipping_method ||
         app.getters.hasSelectedInStore(state))
    },

    /**
     * @param {state}
     * @returns {boolean}
     */
    isEditingEmail: function (state) {
      return state.isEditingEmail
    },

    isEditingAddress: function (state) {
      return state.isEditingAddress
    },

    /** 
     * If a user's email address is found, but the are not 
     * currently logged in.
     * @param {Object} state
     * @returns {boolean}
     */
    isLoggedOutUser: function (state) {
      return state.isLoggedOutUser
    },

    /** 
     * If we're waiting on the server to return user info.
     * @param {Object} state
     * @returns {boolean}
     */
    isLoadingUser: function (state) {
      return state.isLoadingUser
    },


    isOnlySelectingInStore: function (state) {
      return state.isOnlySelectingInStore
    },

    isOrderLoading: function (state) {
      return state.order.id === undefined || state.order.id === null
    },

    isReusingAddress: function (state) {
      return state.reusedAddressId !== null
    },

    isReviewDisplayed: function (state) {
      return app.getters.isBillingDisplayed(state) &&
        app.getters.hasAttachedPayment(state)
    },

    isReviewOrderButtonDisplayed: function (state) {
      if (app.getters.hasAttachedPayment(state)) {
        return false
      }

      return true
    },

    /**
     * @param {Object} store
     * @returns {Boolean}
     */
    isShippingMethodSelected: function (state) {
      return state.order.shipping_method !== ""
    },

    isShippingDisplayed: function (state) {
      return !app.getters.isOrderLoading(state) && app.getters.hasEmailAddress(state)
    },

    isUsingNewAddress: function (state) {
      if (!app.getters.hasAttachedShippingAddress(state)) {
        return false
      }

      return !app.getters.isReusingAddress(state)
    },

    isWaitingForFinish: function (state) {
      return state.isWaitingForFinish
    },

    order: function (state) {
      return state.order
    },

    paymentMethods: function (state) {
      return state.paymentMethods
    },

    pickupDetails: function (state) {
      return state.pickupDetails
    },

    /**
     * When the server returns a credit card error on finalize,
     * and the user is attempting to reuse a card, that UserCard's
     * id is pushed into this array.
     * @param {Object} state
     * @returns {Array<Number>}
     */
    problemCardIds: function (state) {
      return state.problemCardIds
    },

    reusedAddressId: function (state) {
      return state.reusedAddressId
    },

    reusedCardId: function (state) {
      return state.reusedCardId
    },

    shippingAddress: function (state) {
      return state.shippingAddress
    },

    shippingContact: function (state) {
      return state.shippingContact
    },

    shippingMethods: function (state) {
      return state.shippingMethods
    },

    shippingPromoDollars: function (state) {
      var toDollars = Vue.filter("toDollars")
      return toDollars(state.shippingPromoCents)
    },

    STATES: function (state) {
      return state.STATES
    },

    storeLocations: function (state) {
      return state.storeLocations
    },

    toQualifyDollars: function (state) {
      var toDollars = Vue.filter("toDollars")
      var difference = toDollars(state.shippingPromoCents - state.order.subtotal_cents)

      if (difference <= 0) {
        return 0
      }

      return difference
    },

    user: function (state) {
      return state.user
    },

    userAddresses: function (state) {
      return state.userAddresses
    }

  }

  app.getters = getters

})();
```

**Comment by theotherzach** [comment]
Actions.

```javascript
;(function () {
  "use strict"

  var app = window.app

  var actions = {

    ajaxError: function (store, message, responseJSON, type) {
      store.dispatch("AJAX_ERROR", message, responseJSON, type)
    },

    cancelEditShipping: function (store, isInStore) {
      store.dispatch("TOGGLE_EDIT_SHIPPING", false)
      if (isInStore) {
        store.dispatch("REVERT_PICKUP_DETAILS")
        return
      }
      store.dispatch("REVERT_SHIPPING_ADDRESS")
    },

    creditCardFormValid: function (store) {
      if (store.state.isCreditCardFormValid) { return }
      store.dispatch("CREDIT_CARD_FORM_VALID")
    },

    creditCardFormInvalid: function (store) {
      if (!store.state.isCreditCardFormValid) { return }
      store.dispatch("CREDIT_CARD_FORM_INVALID")
    },

    /**
     * Logs the user in and reloads the page when given
     * a matching email and password combination. Otherwise
     * displays an error message.
     * @param{Object} store
     * @param{Object} payload - email and password
     * @return Promise.{(null|Error)}
     */
    logUserIn: function (store, payload) {
      return $.ajax({
          url: "/users/sign_in",
          method: "POST",
          data: { user: payload }
        }).then(function () {
          window.location.reload()
          return null
        }, function (err) {
          var message = "There was a problem logging you in. You may continue as a guest, try again, or reset your password."
          store.dispatch("AJAX_ERROR", message, null, ["login", "Login"])
          return err
        })
    },

    newShipping: function (store) {
      store.dispatch("SET_SHIPPING_ADDRESS", {})
      store.dispatch("TOGGLE_EDIT_SHIPPING", true)
    },

    editShipping: function (store, address) {
      if (address) {
        store.dispatch("SET_SHIPPING_ADDRESS", address)
      }
      store.dispatch("TOGGLE_EDIT_SHIPPING", true)
    },

    handleAddressFormValid: function (store) {
      store.dispatch("TOGGLE_ADDRESS_FORM_VALID", true)
    },

    handleAddressFormInvalid: function (store) {
      store.dispatch("TOGGLE_ADDRESS_FORM_VALID", false)
    },

    getOrder: function (store) {
      app.services.ordersService.get()
      .then(function (data) {
        store.dispatch("SET_ORDER", data)
      }, function (err) {
        store.dispatch("AJAX_ERROR", "Could not get Cart. Please reload the page", err.responseJSON)
      })
    },

    destroyLineItem: function (store, orderLineItem) {
      app.services.orderLineItemsService.destroy(orderLineItem.id)
      .then(function (data) {
        return app.services.ordersService.get()
      }, function (err) {
        store.dispatch("AJAX_ERROR", "Could not remove product", err.responseJSON, ["orderLineItem", "Line Item"])
      })
      .then(function (data) {
        store.dispatch("SET_ORDER", data)
      }, function (err) {
        store.dispatch("AJAX_ERROR", "Product removed but we had trouble updating your cart totals. Please reload the page")
      })
    },


    /**
     * @param {Object} store
     * @param {string} newEmail
     * @returns Promise.{(string|Error)}
     */
    updateGuestEmail: function (store, email) {
      store.dispatch("CLOSE_EDIT_EMAIL")
      store.dispatch("LOADING_USER")

      return app.services.usersService.guest(email)
      .then(function (data) {
        store.dispatch("USER_LOADED")
        store.dispatch("SET_GUEST_EMAIL", email)
        return data
      }, function (err) {
        store.dispatch("USER_LOADED")
        store.dispatch("AJAX_ERROR", null, err.responseJSON, ["registerGuest", "Register Guest"])
        return err
      })
    },

    setReusedCardId: function (store, cardId) {
      store.dispatch("SET_REUSED_CARD_ID", cardId)
    },

    /**
     * @param {Object} store
     * @param {Object} shippingMethod
     * @returns Promise.{(ShippingMethod|Error)}
     */
    setShippingMethod: function (store, shippingMethod) {
      var savedShippingMethod = Object.assign({}, store.state.shippingMethod)
      store.dispatch("SAVE_SHIPPING_METHOD", shippingMethod)
      return app.services.shippingMethodsService.save(shippingMethod.name)
      .then(function (data) {
        return Promise.resolve(data)
      }, function (err) {
        store.dispatch("SAVE_SHIPPING_METHOD_FAILURE", savedShippingMethod)
        store.dispatch("AJAX_ERROR", null, null, ["shippingMethod", "Shipping Method"] )
        return Promise.reject(err)
      })
    },

    setPayment: function (store, cardParams) {
      var cardClone = Object.assign({}, cardParams)
      var ccn = cardClone.credit_card_number
      if (ccn.length > 4) {
        cardClone.last4 = ccn.substr(ccn.length - 4, 4)
      } else {
        cardClone.last4 = ccn
      }

      store.dispatch("TOGGLE_EDIT_CREDIT_CARD", false)
      store.dispatch("SET_CARD", cardClone)
    },

    closeCardForm: function (store) {
      store.dispatch("TOGGLE_EDIT_CREDIT_CARD", false)
    },

    openCardForm: function (store) {
      store.dispatch("CHECKPOINT_CARD", true)
      store.dispatch("TOGGLE_EDIT_CREDIT_CARD", true)
    },

    /**
     * @param {Object} store
     * @param {Number} userAddressId
     * @returns {undefined}
     */
    setReusedAddressId: function (store, userAddressId) {
      store.dispatch("INVALIDATE_SHIPPING_METHOD")
      store.dispatch("SET_REUSED_ADDRESS_ID", userAddressId)
    },

    setUserAddresses: function (store, data) {
      store.dispatch("SET_USER_ADDRESSES", data)
    },

    /**
     * @param {Object} store
     * @param {string} email
     * @returns Promise.{Error}
     */
    checkForRegisteredUser: function (store, email) {

      var payload = {
        email: email
      }

      store.dispatch("LOADING_USER")

      //TODO: Update endpoint to return success / fail
      return app.services.usersService.save(payload).then(function (data) {
        console.warn("Please implement success flow")
        return data
      }, function (err) {
        store.dispatch("USER_LOADED")
        var isUserRegistered = err.status === 409
        if (isUserRegistered) {
          return store.dispatch("USER_IS_REGISTERED")
        }

        return app.actions.updateGuestEmail(store, email)
      })

    },

    persistItem: function (store, itemParams) {
      store.dispatch("SET_ITEM", itemParams)
      app.services.orderLineItemsService.save(itemParams)
      .then(function (data) {
        store.dispatch("SET_ITEM", data)
        return app.services.ordersService.get()
      }, function (err) {
        store.dispatch("REVERT_ITEM", itemParams.id)
        store.dispatch("AJAX_ERROR", null , err.responseJSON, ["orderLineItem", "Line Item"])
      })
      .then(function (data) {
        store.dispatch("SET_ORDER", data)
      }, function (err) {
        store.dispatch("AJAX_ERROR", "Product changed but we had trouble updating your cart totals. Please reload the page", err.responseJSON)
      })
    },

    /**
     * Coordinates network calls, responses, and resulting
     * mutations for finalizing an order.
     * @param {Object} store
     * @returns Promise.{(undefined|Error)}
     */
    placeOrder: function (store) {
      if (store.state.isWaitingForFinish) { return }
      if (!app.getters.isReviewDisplayed(store.state)) { return }

      store.dispatch("TOGGLE_IS_WAITING_FOR_FINISH", true)
      return app.services.ordersService.finish().then(function (data) {
        store.dispatch("TOGGLE_IS_WAITING_FOR_FINISH", false)
        return data
      }, function (err) {
        store.dispatch("TOGGLE_IS_WAITING_FOR_FINISH", false)
        var errorObject = err.responseJSON

        if (err.status !== 422) {
          errorObject = null
        }

        if (errorObject && errorObject.credit_card) {
          if (app.getters.hasEnteredNewCard(store.state)) {
            store.dispatch("CHECKPOINT_CARD", true)
            store.dispatch("TOGGLE_EDIT_CREDIT_CARD", true)
          } else {
            store.dispatch("REUSED_CARD_ERROR")
          }
        }

        store.dispatch("AJAX_ERROR", null, errorObject, ["finishOrder", "Place Order"])

        return err
      })
    },

    /**
     * @param {Object} store
     * @param {Number} userAddressId
     * @returns {undefined}
     */
    updateShippingAddress: function (store, payload) {
      store.dispatch("SET_SHIPPING_ADDRESS", payload)

      if (store.state.reusedAddressId) {
        store.dispatch("SET_REUSED_ADDRESS_ID", null)
      }

      if (store.state.isEditingAddress) {
        store.dispatch("TOGGLE_EDIT_SHIPPING", false)
      }

      if (store.state.order.shipping_method) {
        store.dispatch("INVALIDATE_SHIPPING_METHOD")
      }
    },

    setShippingMethods: function(store, shippingMethods) {
      store.dispatch("SET_SHIPPING_METHODS", shippingMethods)
    },

    updateInStorePickupInfo: function (store, payload) {
      store.dispatch("TOGGLE_EDIT_SHIPPING", false)
      store.dispatch("SET_PICKUP", payload)
    },

    viewInStoreSelect: function (store) {
      store.dispatch("TOGGLE_EDIT_SHIPPING", true)
      store.dispatch("SET_ONLY_IN_STORE")
    },

    checkPromo: function (store, promoCode) {
    },

  }

  app.actions = actions

})();
```

**Comment by theotherzach** [comment]
My store.

```javascript
;(function () {
  "use strict"

  var app = window.app
  var gon = window.gon
  var HARD_CODED_STORE_LOCATIONS = [{ // FIXME: Hard coded
    name: "***",
    id: 1
  }, {
    name: "Option 2",
    id: 2
  }, {
    name: "Option 3",
    id: 3
  }]

  var HARD_CODED_SHIPPING_PROMO_CENTS = 5000 // FIXME: Hard coded

  var store = {

    state: {
      hasActivatedEditCreditCard: null,
      isLoggedOutUser: false,
      isLoadingUser: false,
      isAddressFormValid: false,
      isWaitingForFinish: false,
      isCreditCardFormValid: false,
      isEditingAddress: false,
      isEditingEmail: false,
      isOnlySelectingInStore: false,
      COUNTRIES: app.constants.COUNTRIES,
      user: gon.current_user,
      userAddresses: gon.user_addresses,
      order: gon.order,
      pickupDetails: {},
      shippingContact: {
        first_name: undefined,
        last_name: undefined,
        phone: undefined
      },
      shippingAddress: {},
      reusedAddressId: null,
      creditCard: {},
      paymentMethods: gon.payment_methods,
      reusedCardId: null,
      problemCardIds: [],
      shippingMethods: [],
      shippingPromoCents: HARD_CODED_SHIPPING_PROMO_CENTS,
      STATES: app.constants.STATES,
      storeLocations: HARD_CODED_STORE_LOCATIONS,
      checkpoints: {},
      clones: {
        items: {},
        order: {},
        shippingAddress: {},
        pickupDetails: {}
      },
      errors: []
    },

    middlewares: [{
      onInit: function (state, store) {
        var shippingContactKeys = Object.keys(state.shippingContact)
        shippingContactKeys.forEach(function (key) {
          state.shippingContact[key] = gon.shipping_address[key] || gon.current_user[key]
        })

        var shippingAddress = Object.keys(gon.shipping_address)
        .filter(function (key) {
          return shippingContactKeys.indexOf(key) === -1
        })
        .reduce(function (memo, key) {
          memo[key] = gon.shipping_address[key]
          return memo
        }, {})

        state.shippingAddress = shippingAddress

        state.clones.order = Object.assign({}, state.order)
        state.clones.shippingAddress = Object.assign({}, state.shippingAddress)
        state.clones.pickupDetails = Object.assign({}, state.pickupDetails)

        if (state.order.id) {
          state.reusedAddressId = app.services.shippingAddressesService.reusedId(state.order.id)
          state.reusedCardId = app.services.paymentsService.reusedId(state.order.id)
        }

        state.checkpoints = {
          reusedCardId: state.reusedCardId,
          creditCard: state.creditCard
        }
      }
    }],

    mutations: {

      AJAX_ERROR: function (state, message, err, type) {
        state.errors.push({
          message: message,
          errorPayload: err,
          type: type
        })
      },

      CHECKPOINT_CARD: function (state) {
        state.checkpoints.creditCard = Object.assign({}, state.creditCard)
      },

      CREDIT_CARD_FORM_VALID: function (state) {
        state.isCreditCardFormValid = true
      },

      CREDIT_CARD_FORM_INVALID: function (state) {
        state.isCreditCardFormValid = false
      },

      /**
       * @param {Object} state
       * @returns {undefined}
       */
      INVALIDATE_SHIPPING_METHOD: function (state) {
        state.order.shipping_method = ""
        state.order.shipping_total_cents = 0
      },

      /**
       * @param {Object} state
       * @returns {undefined}
       */
      LOADING_USER: function (state) {
        state.isLoadingUser = true
      },

      /**
       * @param {Object} state
       * @returns {undefined}
       */
      REUSED_CARD_ERROR: function (state) {
        state.problemCardIds.push(state.reusedCardId)
      },

      SET_CARD: function (state, card) {
        state.reusedCardId = null
        state.creditCard = card
      },

      /**
       * @param {Object} state
       * @param {string} newEmail
       * @returns {undefined}
       */
      SET_GUEST_EMAIL: function (state, newEmail) {
        state.order.email = newEmail
      },

      SET_ORDER: function (state, order) {
        state.order = order
      },

      SET_PICKUP: function (state, pickupParams) {
        state.clones.pickupDetails = Object.assign({}, state.pickupDetails)
        var newPickupDetails = Object.assign({}, pickupParams)
        state.shippingContact = {
          first_name: newPickupDetails.first_name,
          last_name: newPickupDetails.last_name,
          phone: newPickupDetails.phone
        }
        delete newPickupDetails.first_name
        delete newPickupDetails.last_name
        delete newPickupDetails.phone
        state.pickupDetails = newPickupDetails
        state.shippingAddress = {}
        state.order.shipping_address_id = null
      },

      SET_REUSED_CARD_ID: function (state, cardId) {
        state.creditCard = {}
        state.reusedCardId = cardId
      },

      NEW_SHIPPING_ADDRESS: function (state) {
        state.clones.shippingAddress = Object.assign({}, state.shippingAddress)
        state.shippingAddress = Object.keys(state.shippingAddress)
        .reduce(function(memo, key) {
          if (key === "id" || key.match(/_id$/)) {
            memo[key] = undefined
            return memo
          }

          memo[key] = ""
          return memo
        }, {})
      },

      /*
       * @param {Object} state
       * @param {ShippingMethod} shippingMethod
       * @returns {undefined}
       */
      SAVE_SHIPPING_METHOD: function (state, shippingMethod) {
        state.order.shipping_total_cents = shippingMethod.cost_cents
        state.order.shipping_method = shippingMethod.name
      },

      /*
       * @param {Object} state
       * @param {ShippingMethod} shippingMethod
       * @returns {undefined}
       */
      SAVE_SHIPPING_METHOD_FAILURE: function (state, originalShippingMethod) {
        state.order.shipping_total_cents = originalShippingMethod.cost_cents
        state.order.shipping_method = originalShippingMethod.name
      },

      SET_SHIPPING_ADDRESS: function (state, addressParams) {
        state.clones.shippingAddress = Object.assign({}, state.shippingAddress)
        var newShippingAddress = Object.assign({}, state.shippingContact, addressParams)
        state.shippingContact = {
          first_name: newShippingAddress.first_name || state.user.first_name,
          last_name: newShippingAddress.last_name || state.user.last_name,
          phone: newShippingAddress.phone
        }
        delete newShippingAddress.first_name
        delete newShippingAddress.last_name
        delete newShippingAddress.phone
        state.shippingAddress = newShippingAddress
        state.pickupDetails = {}
        state.order.shipping_address_id = state.shippingAddress.id
      },

      SET_SHIPPING_METHODS: function(state, shippingMethods) {
        state.shippingMethods = shippingMethods
      },

      SET_USER_ADDRESSES: function (state, addresses) {
        state.userAddresses = addresses
      },

      SET_REUSED_ADDRESS_ID: function (state, userAddressId) {
        state.reusedAddressId = userAddressId
      },

      REVERT_SHIPPING_ADDRESS: function (state) {
        state.shippingAddress = Object.assign({}, state.clones.shippingAddress)
      },

      SET_ITEM: function (state, itemParams) {
        var originalItem = state.order.order_line_items.find(function(item) {
          return item.id === itemParams.id
        })

        // Push the item onto the array and return if originalItem
        // is not found.
        if (!originalItem) {
          state.order.order_line_items.push(itemParams)
          return
        }

        // Create a backup of the original item
        var clonedItem = Object.assign({}, originalItem)
        state.clones.items[clonedItem.id] = clonedItem
        state.clones.items = Object.assign({}, state.clones.items)

        // Replace the original item with an updated copy
        var index = state.order.order_line_items.indexOf(originalItem)
        var updatedItem = Object.assign({}, originalItem, itemParams)
        state.order.order_line_items.$set(index, updatedItem)
      },

      SET_ONLY_IN_STORE: function (state) {
        state.isOnlySelectingInStore = true
      },

      REVERT_ITEM: function (state, id) {
        if (!id) { return }

        var currentItem = state.order.order_line_items.find(function(item) {
          return item.id === id
        })

        // Return early if we cannot find the current item
        if (!currentItem) {
          return
        }

        // Replace the current item with a clone of the backup
        var previousItem = state.clones.items[currentItem.id]
        var clonedPreviousItem = Object.assign({}, previousItem)
        var index = state.order.order_line_items.indexOf(currentItem)
        state.order.order_line_items.$set(index, clonedPreviousItem)
      },

      REVERT_PICKUP_DETAILS: function (state) {
        state.pickupDetails = Object.assign({}, state.clones.pickupDetails)
      },

      /**
       * @param {Object} state
       * @returns {undefined}
       */
      USER_IS_REGISTERED: function (state) {
        state.isLoggedOutUser = true
      },

      /**
       * @param {Object} state
       * @returns {undefined}
       */
      USER_IS_GUEST: function (state) {
        state.isLoggedOutUser = false
      },

      /**
       * @param {Object} state
       * @returns {undefined}
       */
      USER_LOADED: function (state) {
        state.isLoadingUser = false
      },

      TOGGLE_ADDRESS_FORM_VALID: function (state, isValid) {
        if (isValid === undefined) {
          state.isAddressFormValid = !state.isAddressFormValid
          return
        }

        state.isAddressFormValid = isValid
      },

      TOGGLE_IS_WAITING_FOR_FINISH: function (state, isWaiting) {
        if (isWaiting === undefined) {
          state.isWaitingForFinish = !state.isWaitingForFinish
          return
        }

        state.isWaitingForFinish = isWaiting
      },

      TOGGLE_EDIT_CREDIT_CARD: function (state, shouldEdit) {
        if (shouldEdit === undefined) {
          state.hasActivatedEditCreditCard = !state.hasActivatedEditCreditCard
          return
        }

        state.hasActivatedEditCreditCard = shouldEdit
      },

      TOGGLE_EDIT_SHIPPING: function(state, shouldEdit) {
        if (shouldEdit === undefined) {
          state.isEditingAddress = !state.isEditingAddress
        } else {
          state.isEditingAddress = shouldEdit
        }

        if (!state.isEditingAddress) {
          state.isOnlySelectingInStore = false
        }

      },

    }

  }

  app.store = store

})();
```

**Comment by theotherzach** [comment]
Note the inconsistent and flailing nature of my Vuex use. I'm looking at this (now public) code and I'm deeply unhappy with it. I'm going to delete these specific code comments in a day or two so they don't become too Google-able, but I wanted to be fully transparent when I'm saying that I'm having trouble figuring out how to properly use this library.

**Comment by theotherzach** [comment]
## Proposal:

### Deprecate mutations and getters.

* Spend some time with Redux's [original inspiration](https://reagent-project.github.io/index.html) to see how nice life is without needing to explicitly declare getters and setters.
* Vuex Mutations: - deprecated
* Vuex Getters: - deprecated, or change their name to Vuex Computed
* Vuex State: - no change
* Component: Vuex state read access - 
  * If there is a single module, all components get access to the entire state tree. (for smaller apps)
  * If there are modules, then state is exposed through granting access at the module level. `Vuex.mapModule`
  * Which might mean that auto-namespacing modules makes sense again.
  * If restricting access to read-only state is a concern, then a variation of that could be `Vuex.stateKeys`, or leave both options open to see if one is a foot-gun, like event propagation. 
* Component: Vuex state write access - none
* Vuex Actions - Now directly mutate the store, but maybe through an explicit `COMMIT` function or method. 

### Background

Sorry for all my text in this issue today, but I'm extracting methods into actions right now and I'm thinking of the subject a lot. I think the real issue I'm having is the ceremony of the mutations and the getters. That might be because I'm thinking back to my days building UI's using ClojureScript's Reagent (A React wrapper) back in late 2014. I purposely left the Clojure community because I couldn't get much work done with Clojure's `nil` handling, but I do miss how Reagent handled state. 


**Static Component**

Just a function that returns a data structure which conforms to Hiccup-like standards. 

```Clojure
(defn simple-component []
  [:div
   [:p "I am a component!"]
   [:p.someclass
    "I have " [:strong "bold"]
    [:span {:style {:color "red"}} " and red "] "text."]])
```

**Props**

Note that props aren't declared ahead of time, just passed in and used. (I'm not proposing that Vue adopts this I <3 prop type checking, but it might be an important detail.)

```Clojure
(defn hello-component [name]
  [:p "Hello, " name "!"])

(defn say-hello []
  [hello-component "world"])
```

**State**

State is handled by updating a [Clojure Atom](http://clojure.org/reference/atoms), which is a special type of Clojure reference. Briefly, all Clojure data-structures are made persistent (or immutable) through structural sharing. When writing Clojure, if you want to deal in the filthy filthy state, you use a special construct such as an atom. To set the state of an atom, you use a specific function such as `swap!`. If you want to get the current state of the atom you "deref" it by appending `@` to your reference to the atom. `@my-cool-atom`. 

```Clojure
(defn timer-component []
  (let [seconds-elapsed (r/atom 0)]
    (fn []
      (js/setTimeout #(swap! seconds-elapsed inc) 1000)
      [:div
       "Seconds Elapsed: " @seconds-elapsed])))
```


Back in late 2014, we started off by using multiple atoms on a single page, roughly equivalent to one for each mutation / getter combo in Vuex. As time went on, we started to notice that if we used fewer atoms, our code actually got easier to reason about and change was easier to track. We eventually took this to the logical conclusion and started putting all of our application state in a single atom that we would initialize when the app booted and then passed that atom down into child components via props. 

It was glorious except for a few things:

* Indexing into a deeply nested data structure in Clojure is nasty. I wish we had [Specter](http://nathanmarz.com/blog/functional-navigational-programming-in-clojurescript-with-sp.html) back then.
* Performance.
* We still needed to pass the single atom down through props so that the components knew when to update. 

Why am I talking about ClojureScript? Because my understanding is that some Elm patterns were inspired by ClojureScript + React state management and Redux was inspired by Elm patterns. We had no freaking clue at the time, but we were using an application store. We were just using an application store without needing to declare getters and setters which is really nice.

**Comment by fnlctrl** [comment]
~~Congratulations! You sir, just took self-importance to the next level!~~

**Comment by theotherzach** [comment]
I'm not sure how me posting my own poorly written code, despite how badly I feel about it, is self importance. I don't feel important. I feel like maybe I'm not as good a coder as I thought I was because I can't seem to make Vuex work for me in a low friction way.

**Comment by karol-f** [comment]
@theotherzach As I understand, in this discussion on new 2.0 design, new propositions are warmly welcomed but not **noise**, which can be triggered by to many comments. 

I don't want to write in @fnlctrl name but maybe you could just combine your comments into **one** comment and put code examples in external JSFiddle, providing just link. This would highly increase readability of this post and hopefully help in designing better vuex@2.0 :)

**Comment by posva** [comment]
@theotherzach Can you please combine all your code on a jsfiddle and remove the comments as @karol-f said, please? That'd improve the readability of the thread 😄

**Comment by piknik** [comment]
@theotherzach Nice misuse of Github. You really should post this code in the vue.js forums or make a new issue, not put this in a discussion for the future of Vuex, and try to make suggestions when you have not even grasped the current fundamentals! I'm really trying to hold back negative comments, please keep in mind for the future that posting (irrelevant content) many times can annoy some of the subscribers of the issue (email goes straight to my phone).

Please do what @posva said to do.

**Comment by theotherzach** [comment]
I apologize. You won't see me here again.

**Comment by posva** [comment]
@theotherzach I'd appreciate you still share that code in a proper env like jsfiddle 😅 
No need to leave 😉

**Comment by thoughts1053** [comment]
I love everything about the proposal and new features and will be moving to this version as soon as I get some time to refactor.

My only question is regarding module getters receiving the sub state tree: I currently normalize all my api requests and put the different entities in different modules. This will cause all my getters to live on the root store when I want to denormalize an item. Any reason why the module getters can't receive the full state? Is it just to align with the module mutations?

**Comment by yyx990803** [comment]
@theotherzach long snippets are more appropriate in a [gist](https://gist.github.com/), and a proposal, especially one that directly contradicts this design, would be more fitting as a separate issue.

That said, I think your usage (from my memory of glancing over what you posted before) were indeed missing a few points about Vuex:

1. As you've noticed, managing state through Vuex does add additional boilerplate, so you need to be aware of the tradeoff. The benefits of Vuex is that changes going through the store are **trackable**, **replayable** and **restorable**. This is particularly useful when combined with the devtool to help you understand what is going on with your application state. The design is also focused on providing a scalable structure for large applications with multiple teams - the "boilerplate" part is there to ensure everyone with knowledge of how Vuex works knows where to look for things when they jump into a Vuex-based app.

2. It's not mandatory, nor recommended to put **all your state** inside a Vuex store. The examples are probably to be blamed because they are contrived and in fact too trivial to require using Vuex (they just examples). The guide mentions upfront that it's important to distinguish between **application-level state** and **component local state**. If you are not sure whether a piece of state belongs to a component or the store, start with local state. For example, a temporary flag like `isLoadingUser` should most likely be component local state. In most cases, you'd extract state into the store for two purposes:

  1. You realize it needs to be shared among multiple components in the entire application.
  2. It is important enough that you want to require dispatching actions to change it, so that you can track it in devtools.

  If you simply put every single property in a big component into Vuex, you will most likely end up with more code and no obvious gain.

3. Getters are not mandatory. In fact their purpose is for computing **derived state**. Nothing prevents you from directly accessing store state, even in templates you can do `{{ $store.state.someData }}`.

4. The distinction between mutations and actions is important. An action stand for a possibly asynchronous sequence of operations, while mutations stand for a synchronous, trackable transaction. The guide already explained this.

Finally, if you don't think it's worth the extra code to gain trackability of your state (as shown from your proposal), you are probably better off just stay with component local state + [the simple store pattern mentioned here](http://vuejs.org/guide/application.html#State-Management).

**Comment by theotherzach** [comment]
@posva Sorry for my over-reaction but I'm late on this project and exhausted. In short I find the following based on my experiences. I hope that I'm using Vuex wrong, but even if that's the case then perhaps this points to evidence that Vuex is easy to use wrong.

* Vuex is great for me when I know ahead of time that I've got enough state to warrant it and I don't have too many async calls.
* The more async calls I have, the more mutation & getter boilerplate I write.
  * Loading indicators
  * recovering after a network call failed after I optimistically updated the UI, such as re-opening a previously closed address form and highlighting the fields with errors.
* I find it very difficult to keep state in both components and in the vuex store. 

https://gist.github.com/theotherzach/92ddb73af8f9e4c8ab70ca82a49c8fc1

**Comment by yyx990803** [comment]
@thoughts1053 I think it's probably a good idea to expose an extra `rootState` for module getters/mutations/actions.

**Comment by thoughts1053** [comment]
@yyx990803, hopefully last questions regarding getters:

1. Are getters able to currently call other getters from different modules? - I guess we can just export the getter and manually import in in another module.

2. Can getters take arguments?

- Regarding denormalized data (ie, posts and comments) and using getters from different modules, this would allow us to do something like `getCommentsForPost(postId)`.

- Regarding performance (and this may not be an issue), currently I have to get a list of posts (keeping with the above example) and pass each post as a prop into a component. When updating a single post in the list, is the list of posts retrieved again and then each component re-rendered? If so, wouldn't having the component have a getter `getPost(postId)` prevent unnecessary re-renders of other components?

**Comment by piknik** [comment]
@thoughts1053 You hit the nail on the head on both points. I find myself doing _1_ a lot, and _2_ would make life so much easier.

Currently in 1.0 when you call a getter from a getter, it requires passing in the state/store to the next getter, and putting each getter into an object that can be accessed by all getters. Would be nice to give each getter some way of calling another getter like how you would in a component `this.cartItems` vs `getters.cartItems(store)`. Even nicer to provide arguments, example `this.cartItems(accountId)`.

**Comment by rhan-mentad** [comment]
@yyx990803
I think it would be helpful to have default mutators/actions that perform a "set" behavior and thus not require explicit boilerplate declarations. Having to write custom get/set code tends to be the exception, not the rule, imho. This provides the benefits of track ability while reducing the amount of code needed.

**Comment by LinusBorg** [comment]
@thoughts1053 

**Concerning 1.:** 
The are now in [2.0.0.-rc3](https://github.com/vuejs/vuex/releases/tag/v2.0.0-rc.3)

**Concerning 2.:** 
I don't think this makes a lot of sense. to me, there are two scenarios where a parameter might make some sense:

1. you have some local state and want to filter depending on that. in that case, you could simply use a computed property. getters should be pure functions to be re-usable independently from component state, so they cannot access local state like a compute prop in a component can.

A getter in the store would make not much sense to me anyway, as the operation is only meant for that component anyway.

2. you have a collction in the store (i.e. `comments`) as well as the filter criterion; `currentPost`. In that situation you don't need to receive the `currentPost` as a parameter because it is already available in the store, especially when now you can use other getters in a getter.

**Comment by thoughts1053** [comment]
@LinusBorg Thanks! I saw the release last night.

You've also answered both my questions regarding passing parameters to getters. #1 was my main concern, but creating a local computed property with `this.$store` makes sense. Thanks again.

**Comment by LinusBorg** [comment]
@theotherzach 

Looking at the gist you posted, that should probably be split up into some state submodules to make it more maintanable.

concerning the boilerplate when having async actions: The 2.0 design proposal actually makes this much easier since async actions now should return a Promise:

```
this.loading = true
store.dispatch('doSomething', { id: 1 }).then(() => {
  this.loading = false
})
.catch(function(err){
  this.loading = false
  // call this async action again, and // or handle error display ...
})
```

**Comment by analog-nico** [comment]
I just upgraded to `vuex@2.0.0-rc.3` and as a result have two small suggestions:

1. For convenience `mapActions` and `mapGetters` should map ALL available actions/getters if they are called without any parameters.
2. Suggesting to use `mapActions` and `mapGetters` with the Object spread operator is very neat but I only realized that the Object spread operator is part of the ES7 draft spec after I ran into many issues (eslint does not support it by default, buble requires a polyfill). I suggest adding the following sentence to the documentation: "The Object spread operator is part of the ES7 specification. Please verify that your toolchain is properly configured to support this operator."

**Comment by karol-f** [comment]
@analog-nico #2 I only had to use this - https://babeljs.io/docs/plugins/transform-object-rest-spread/. No other setup was needed in my case. I agree however about the need of info in documentation.

**Comment by NicolasParada** [comment]
I didn't test it yet, so... does Object.assign() work?

```js
computed: Object.assign({}, 
  localComputed () { … },
  mapGetters(['a', 'b', 'c', 'd'])
),
methods: Object.assign({},
  localMethod () { … },
  mapActions(['b'])
)
```

??

**Comment by analog-nico** [comment]
@karol-f True, as you say, the note in the docs is exactly my point. Likely any toolchain will easily support the operator. But since it is ES7, chances are, it's not supported by default. So it is better having the note than letting more people like me turning their toolchain inside out. ;)

@NicolasParada `Object.assign` is a good ES6 alternative. Your code works if you add curly braces around `localComputed` and `localMethod`. However, `Object.assign` needs a polyfill to run in ES5 environments. Babel likely adds one automatically. For buble you have to add it yourself.

**Comment by NicolasParada** [comment]
@analog-nico Ah, of course, I forgot them.

**Comment by Namek** [comment]
@yyx990803 what's the recommended way regarding getters with parameters in Vuex 2.0? I used them as helpers, e.g. for some calendar I have `v-for` looping through `days` and getting them using `getDay()` helper as such:

```
function getDayIndex(weekIndex, weekDayIndex) {
  return weekIndex*c.WEEK_DAY_COUNT + weekDayIndex
}

export const getDay = ({calendar}) => (weekIndex, weekDayIndex) => {
  return calendar.days[getDayIndex(weekIndex, weekDayIndex)]
}
```

**Comment by ronwong82** [comment]
@yyx990803 I get error dispatch undefined. FYI I'm using vuex 2.3. what's wrong?

**Comment by LinusBorg** [comment]
This is not a support forum. Please ask questions on forum.vuejs.org

Also, please provide code that we and other people on the forum can analyse to solve your problem.
